pkg/webhook/resources/machine/validator.go

```diff
@@ -37,6 +45,28 @@ func (v *machineValidator) Resource() types.Resource {
 func (v *machineValidator) Create(req *types.Request, newObj runtime.Object) error {
 
 	// if cluster is upgrading, intercept the request
+	isUpgrading, err := v.isClusterUpgrading()
+	if err != nil {
+		return fmt.Errorf("failed to check if cluster is upgrading: %w", err)
+	}
+	if isUpgrading {
+		return fmt.Errorf("cluster is upgrading, cannot create machine")
+	}
 
 	return nil
 }
+
+func (v *machineValidator) isClusterUpgrading() (bool, error) {
+
+	upgrades, err := v.upgradeCache.List(util.HarvesterSystemNamespaceName, labels.NewSelector())
+
+	if err != nil {
+		return false, err
+	}
+	for _, upgrade := range upgrades {
+		if !harvesterv1.UpgradeCompleted.IsFalse(upgrade) {
+			return true, nil
+		}
+	}
+	return false, nil
+}
```

Comment:
IIUC, `isClusterUpgrading()` would return `true` even for an Upgrade that failed to complete in the past. Might be better off checking only the last upgrade in the `upgrades`, and check if it's in `Unknown` state because that, AFAIK, is the indicator of an upgrade in progress.

And, of course, I'm not sure if this is the right resource. My comment is only for the logic. :)

Reply:
Yes, if any Upgrade CRs have a `Completed` condition of `Unknown`, it indicates an ongoing upgrade. Ideally, there should be only one Upgrade CR that is `Unknown` in the `Completed` condition.